postmortem/2021-08-06-member-invitation-leak.md

```diff
@@ -0,0 +1,61 @@
+
+Following the specifications for the new refactor, a new query interface was implemented for fetching MemberInvitations and naturally, the first implementation submitted by the user was carrying the former abstraction: its arguments were CollectiveId and MemberCollectiveId, following the abstractions existing in our database model. This was instantly caught in review and fixed with the new GraphQL V2 abstractions: Account and MemberAccount.
+
+Although the code was reviewed multiple times by core team members, one error slipped: the query executed by this resolver was still searching for the arguments CollectiveId and MemberCollectiveId and since the input arguments were updated, these would always be undefined, and since they were undefined, the MemberInvitationQuery that should fetch MemberInvitations for a specific account would return all MemberInvitations in the DB.
+
+This error was originally caught by the author after the code was deployed to production. The author opened a PR and submitted for the responsible core team member to review. The responsible core team member usually reads his notifications once as the first action of the day and since this was submitted after, it was bound to be caught on the next day.
+
+The core team was notified about the issue by an email sent to our security email by an user that was worried this was a critical bug that would allow them to join any collective.
+This was fortunately not the case, we were just wrongly displaying all open invitations as if it was meant to the logged in user.
+
+## Impact
+
+From 12:24PM to 7:40PM UTC, users were seeing all Member Invitations in the Pending Invitations page as if they were invited to those collectives.
+
+## Timeline
+
+(UTC August 5th, 2021)
+
+- **12:24**
+  Faulty PR is deployed to production.
+- **13:21**
+  Contributor submits patch PR for review.
+- **19:31**
+  Security email is posted on slack.
+- **19:40**
+  Contributor PR is deployed to production.
+
+## Root Cause
+
+Faulty refactor in the original PR: https://github.com/opencollective/opencollective-api/pull/6008
+
+## Resolution
+
+Fixed variables in patch 6408: https://github.com/opencollective/opencollective-api/pull/6408
+
+## What went well?
+
+- The review caught the inconsistency in the variable names.
+- A patch was promptly submitted by the contributor.
+
+## What could have been better
+
+- Our existing resolvers should by default implement typing so contributors can take them as example.
+- Writing more tests and request that from contributors.
+- We could DM responsibles as a way to ensure they're aware of important patches even though we notified them in GitHub.
```

Comment:
I have a suggestion on this, especially on the third point. 💡 

This DM process might be useful but I think we need to put it in our docs so that contributors will know this expectation. My suggestion is to add a line here; https://docs.opencollective.com/help/contributing/development/best-practice-guidelines or maybe https://docs.opencollective.com/help/about/contributing#code-contributions along the lines of; 

>If your PR is merged and later you found a bug/issue with the code that you've submitted, please DM a core team member on Slack. This would be really helpful for us so that we could fix potentially harmful issues quickly. 

It this is in agreement I'll add it to the dos. 🙂

Reply:
I like that!

Reply:
DM is not a good practice, information is sent to a single person and not shared with the team.

Reply:
Maybe email to support@opencollective.com ?

Reply:
Email is kinda useless for me. I rather have someone communicating straight through Slack if it is urgent.

Reply:
Maybe we can create a dedicated channel for this. 🤔

Reply:
We generally have the #engineering channel that should be good enough for most cases.